consensus/ethash/consensus.go

```diff
@@ -275,11 +275,18 @@ func (ethash *Ethash) verifyHeader(chain consensus.ChainHeaderReader, header, pa
 	if header.GasLimit > cap {
 		return fmt.Errorf("invalid gasLimit: have %v, max %v", header.GasLimit, cap)
 	}
-	// Verify that the gasUsed is <= gasLimit
-	if header.GasUsed > header.GasLimit {
-		return fmt.Errorf("invalid gasUsed: have %d, gasLimit %d", header.GasUsed, header.GasLimit)
-	}
 
+	// Verify the block's gas usage and (if applicable) verify the base fee.
+	if chain.Config().IsAleut(header.Number) {
+		if err := misc.VerifyEip1559Header(parent, header, chain.Config().IsAleut(parent.Number)); err != nil {
+			return err
+		}
+	} else {
+		// Verify that the gasUsed is <= gasLimit
+		if header.GasUsed > header.GasLimit {
+			return fmt.Errorf("invalid gasUsed: have %d, gasLimit %d", header.GasUsed, header.GasLimit)
+		}
+	}
```

Comment:
This is a consensus-flaw, if someone constructs (and mines) a a block where the header contains a basefee, _before_ 1559 is activated. Geth will marshal it properly, and then it will enter here, and we'll accept the block. 

This clause needs to verify that `baseFee` is `nil`. Or, perhaps, check some accessor like `!header.Is1559()`.

Reply:
Yep, you're right. I'll fix this.

Reply:
Actually, this would be stopped by logic in `state_transition.go`: https://github.com/quilt/go-ethereum/blob/48998a9c839a6136d00c5a57cc042ca5ed4ba9e8/core/state_transition.go#L227

Do you think this check should be in both places or one place? And if one place, which of the two should it be in?

Reply:
So, it definitely _needs_ to be checked here. State transition only comes into play when we sync block by block, either `syncmode=full` or we're already in sync. But fast, snap, and light just checks the headers, so the ethash/consensus and clique/consensus need to verify the header correctness fully. 

Whether it needs to be checked in state_transition, however... I would say "Probably not". The only case for having it in state_transition, that I can think of, would be to have the check in place when we execute state-tests, or `evm t8n`, or any test which just uses a prestate and applies a message. But I don't think the actual chain progression could ever hit those errors. So I dunno, leaving them in is kind of 'defensive'. I think maybe we should remove those checks (?)

Reply:
Ah I see!

I am okay to remove them, but I think we need a way to deal with `t8n`. Without the checks here, you could define a `baseFee` in the `env` and run it on a pre-1559 config without error. IUC, we would want those things to be errors?

Reply:
Been thinking about this. So, `t8n` is a state transition tool. It does not handle blocks, or headers. It's just a state + environment variables, and the given ruleset, and go. 

So I think it's fine if `baseFee` _is_ defined in the env, when running `t8n`. It should simply ignore it.